org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SaveActionHandler.java

```diff
@@ -155,4 +174,42 @@ private List<TextEdit> handleSaveActionOrganizeImports(String documentUri, IProg
 		return edit;
 	}
 
+	private void handleSaveActionRenameFile(String documentUri, IProgressMonitor monitor) {
```

Comment:
This is all nice in terms of API usage. You get the name from the problem marker, as well as the offsets. You then use all of this to generate a problem location, which when passed to  `ReorgCorrectionsSubProcessor.getWrongTypeNameProposals`  will give you a proposal that does everything for you. You then convert that proposal to a workspace edit. Nothing is really done "manually", but in this case I wonder if that's the best approach.

From the moment we detect `PublicClassMustMatchFileName`, why can't we immediately get the element name of the `ICompilationUnit` and "manually" create a `WorkspaceEdit` that will rename the file (we have the original `documentUri` and the new name, `${newCompilationUnitName}.java`) ? Are there any corner cases that passing it through `getWrongTypeNameProposals` might do beyond just "renaming the file" ?

Reply:
I had this thought as well, though I wasn't sure it wasn't "hacky" to do it manually. Not sure about the corner cases but I opted to do it this way because it felt safer. I thought there might be some more complex cases when it came to refactoring. I can update and play around with it if you think it would be more appropriate to do it manually

Reply:
If the only thing you have to worry about is renaming the file (if the error exists), that is easy enough to do manually, and should be much less code to manage. It might even be less prone to failure as you're relying on fewer things (no need to access problem for name/offsets, then linking it to the refactoring proposal)